mlflow/gateway/providers/mlflow.py

```diff
@@ -0,0 +1,136 @@
+        self.mlflow_config: MLflowConfig = config.model.config
+        self.headers = {"Content-Type": "application/json"}
+
+    @staticmethod
+    def process_payload(payload, key):
+        payload = jsonable_encoder(payload, exclude_none=True)
+
+        # Remove the route-added regulation value from being passed to MLflow hosted models
+        payload.pop("candidate_count", None)
+
+        input_data = payload.pop(key)
+        request_payload = {"inputs": input_data if isinstance(input_data, list) else [input_data]}
+
+        if payload:
+            request_payload["params"] = payload
+
+        return request_payload
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        # Example request to MLflow REST API server for completions:
+        # {
+        #     "inputs": ["hi", "hello", "bye"],
+        #     "params": {
+        #         "temperature": 0.5,
+        #         "top_k": 3,
+        #     }
+        # }
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_api_base,
+            path="invocations",
+            payload=self.process_payload(payload, "prompt"),
+        )
+
+        # Example response:
+        # {"predictions": ["hello", "hi", "goodbye"]}
+
+        return completions.ResponsePayload(
+            **{
+                "candidates": [
+                    {"text": resp["predictions"], "metadata": {"finish_reason": "length"}}
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        # Example request to MLflow REST API for chat:
+        # {
+        #     "inputs": ["question"],
+        #     "params": ["temperature": 0.2],
+        # }
+
+        payload = self.process_payload(payload, "messages")
+
+        query_count = len(payload["inputs"])
+        if query_count > 1:
+            raise HTTPException(
+                status_code=422,
+                detail="MLflow chat models are only capable of processing a single query at a "
+                f"time. The request submitted consists of {query_count} queries.",
+            )
+
+        payload["inputs"] = [payload["inputs"][0]["content"]]
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_api_base,
+            path="invocations",
+            payload=payload,
+        )
+
+        # Example response:
+        # {"predictions": "answer"}
+
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {"role": "assistant", "content": resp["predictions"]},
+                        "metadata": {"finish_reason": "length"},
```

Comment:
We don't know that the finish reason is length, right? Can we make the `metadata` field an empty dict?